### PR TITLE
[CFM-1030] -Invite member group cta.

### PR DESCRIPTION
**Pull request by ybaras**
https://issuetracker.amplexor.com/jira/browse/CFM-1030

### CODE REVIEW

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module [hunk 1]
```diff
@@ -542,6 +556,18 @@ function c4m_user_og_c4m_join_cta($group) {
 }
 
 /**
+ * Sets the invite members link CTA.
+ */
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module]: @ybaras 
docs are missing ...

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module [hunk 2]
```diff
@@ -915,3 +941,27 @@ function c4m_user_og_user_update(&$edit, $account, $category) {
   // Destroy the entry of this account.
   c4m_user_og_delete_unified_workflow_entries($account->uid);
 }
+
+/**
+ * Return true if the user is allowed to invite and false if he isn't.
+ */
+function check_invite_access_by_group_type($gid, $group_type = 'node') {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module]: @ybaras 
The function should have the prefix of the module.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module [hunk 3]
```diff
@@ -944,14 +947,19 @@ function c4m_user_og_user_update(&$edit, $account, $category) {
 
 /**
  * Return true if the user is allowed to invite and false if he isn't.
+ *
+ * Checks by user role and og_role against group visibility (Public/Private) and
+ * Membership requests setting (Open/Moderated).
  */
-function check_invite_access_by_group_type($gid, $group_type = 'node') {
+function c4m_user_og_check_invite_access_by_group_type($gid, $group_type = 'node') {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module]: @ybaras 
We also discussed about that the `access` should be the suffix

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module [hunk 4]
```diff
@@ -494,6 +504,10 @@ function c4m_user_og_c4m_cta_block() {
     $items['leave'] = $item;
   }
 
+  if (c4m_user_og_invite_access($group->nid)) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module]: @ybaras 
please add a comment for this one

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module [hunk 5]
```diff
@@ -915,3 +944,33 @@ function c4m_user_og_user_update(&$edit, $account, $category) {
...
+
+  // Get group.
+  $group = entity_load_single($group_type, $gid);
+  // Get request type (open/moderated).
+  $membership_request_value = c4m_og_get_group_membership_request($group);
+  // Get access type (public/private).
+  $access_type = c4m_og_get_access_type($group);
+
+  if ($access_type['type'] == 'private' || $membership_request_value == 'moderated') {
+    // Group is either private, restricted-moderated or public-moderated.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module]: move the comment above the condition.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module [hunk 6]
```diff
@@ -915,3 +944,33 @@ function c4m_user_og_user_update(&$edit, $account, $category) {
...
+ *
+ * Checks by user role and og_role against group visibility (Public/Private) and
+ * Membership requests setting (Open/Moderated).
+ */
+function c4m_user_og_invite_access($gid, $group_type = 'node') {
+  global $user;
+  $og_roles = og_get_user_roles($group_type, $gid);
+
+  // Global and Group administrators can invite.
+  if (in_array('administrator', $user->roles) || in_array('administrator member', $og_roles)) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module]: please use `c4m_user_og_get_role_in_group()` and `user_access('administer site') in order to check SA/GO/GA

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module [hunk 7]
```diff
@@ -915,3 +944,33 @@ function c4m_user_og_user_update(&$edit, $account, $category) {
   // Destroy the entry of this account.
   c4m_user_og_delete_unified_workflow_entries($account->uid);
 }
+
+/**
+ * Return true if the user is allowed to invite and false if he isn't.
+ *
+ * Checks by user role and og_role against group visibility (Public/Private) and
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module]: Fix the docs

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module [hunk 8]
```diff
@@ -915,3 +946,34 @@ function c4m_user_og_user_update(&$edit, $account, $category) {
...
+ */
+function c4m_user_og_invite_access($gid, $group_type = 'node') {
+  global $user;
+  $og_roles = og_get_user_roles($group_type, $gid);
+
+  // Get group.
+  $group = entity_load_single($group_type, $gid);
+
+  // SAs and GAs can invite new members.
+  if (c4m_user_og_get_role_in_group($user, $group) == 'administrator' || user_access('administer site')) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module]: @ybaras 
I think that the group owner should also get access in here

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module [hunk 9]
```diff
@@ -915,3 +946,35 @@ function c4m_user_og_user_update(&$edit, $account, $category) {
...
+
+/**
+ * Check current user permission to invite new members to the group.
+ *
+ * SAs and GAs can always invite to the group. GMs can only invite to Open
+ * groups (either public or restricted) but not to moderated or private groups.
+ */
+function c4m_user_og_invite_access($gid, $group_type = 'node') {
+  global $user;
+  $og_roles = og_get_user_roles($group_type, $gid);
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module]: remove this line, since it is not used anymore

**Comment by ybaras** [comment]
**Wide (1280px):**

deskyop

**Mid (960px):**



**Small (480px):**

**Comment by ybaras** [comment]
**Open:**

join

**Moderated:**

request

File: project/profiles/capacity4more/behat/features/group_dashboard.feature [hunk 10]
```diff
@@ -40,3 +40,15 @@ Feature: Group dashboard
     And   I am logged in as user "president"
     And   I visit the dashboard of group "Movie Popcorn Corner"
     Then  I should see the group dashboard without quick post form
+
+  @api
+  Scenario: Check Invite a member link as group owner
+    Given I am logged in as user "mariecurie"
+    When  I visit the dashboard of group "Movie Popcorn Corner"
+    Then  I should see the link "Invite a member"
```
Comment (ordavidil) [inline project/profiles/capacity4more/behat/features/group_dashboard.feature]: @ybaras 
good work with the tests !

can you change it a little bit so the test will click the button and make sure we got the the invite form ?
thanks

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module [hunk 11]
```diff
@@ -915,3 +946,40 @@ function c4m_user_og_user_update(&$edit, $account, $category) {
...
+  // Get request type (open/moderated).
+  $membership_request_value = c4m_og_get_group_membership_request($group);
+  // Get access type (public/private).
+  $access_type = c4m_og_get_access_type($group);
+
+  // If the group is private or moderated, check for power member.
+  // Else just check for member.
+  if ($access_type['type'] == 'private' || $membership_request_value == 'moderated') {
+    return _c4m_features_og_members_is_power_user($group);
+  } else {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module]: @ybaras 
please fix that by coding standards, thanks

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module [hunk 12]
```diff
@@ -970,22 +970,16 @@ function c4m_user_og_invite_access($group = NULL) {
...
   $membership_request_value = c4m_og_get_group_membership_request($group);
   // Get access type (public/private).
   $access_type = c4m_og_get_access_type($group);
 
-  // If the group is private or moderated, regular GMs can't invite.
+  // If the group is private or moderated, check for power member.
+  // Else just check for member.
   if ($access_type['type'] == 'private' || $membership_request_value == 'moderated') {
-    return FALSE;
+    return _c4m_features_og_members_is_power_user($group);
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module]: @ybaras 
it is better to leave that check where it was ...

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module [hunk 13]
```diff
@@ -915,3 +946,45 @@ function c4m_user_og_user_update(&$edit, $account, $category) {
...
+  // Get request type (open/moderated).
+  $membership_request_value = c4m_og_get_group_membership_request($group);
+  // Get access type (public/private).
+  $access_type = c4m_og_get_access_type($group);
+
+  // For private or moderated groups, return false. Else check for member.
+  if ($access_type['type'] == 'private' || $membership_request_value == 'moderated') {
+    return FALSE;
+  }
+  else {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_og/c4m_user_og.module]: @ybaras 
The else is not needed in here ...

**Pull request merged**